Hardy Barth cPμ2 läst sich nicht einbinden - segmentation violation

### Discussed in https://github.com/evcc-io/evcc/discussions/8024

<div type='discussions-op-text'>

<sup>Originally posted by **TobiasHuber1980** May 15, 2023</sup>
Hallo zusammen,

ich bekomme meine Wallbox cPμ2 von Hardy Barth einfach nicht eingebunden.
Hier meine Konfig:

```
yaml
# open evcc at http://evcc.local:7070
network:
  schema: http
  host: evcc.local # .local suffix announces the hostname on MDNS
  port: 7070

log: debug
levels:
  cache: error

# unique installation id
plant: *****

interval: 10s # control cycle interval

sponsortoken: *****

# sponsors can set telemetry: true to enable anonymous data aggregation
# see https://github.com/evcc-io/evcc/discussions/4554
telemetry: false

meters:
- type: template
  template: rct-power
  usage: grid
  host: 192.168.178.10
  name: grid1
- type: template
  template: rct-power
  usage: pv
  host: 192.168.178.10
  name: pv2
- type: template
  template: rct-power
  usage: battery
  host: 192.168.178.10
  name: battery3

chargers:
- type: template
  template: hardybarth-salia
  host: 192.168.178.11
  name: wallbox4

loadpoints:
- title: Garage
  charger: wallbox4
  mode: pv
  phases: 3
  mincurrent: 6
  maxcurrent: 16
  resetOnDisconnect: false

site:
  title: Zuhause
  meters:
    grid: grid1
    pv:
    - pv2
    battery:
    - battery3

</details>
Version: `0.117.3`
```

Fehler beim Erstellen der Konfig:
```
Choose a wallbox: echarge cPH2, cPμ2
-- Sponsorship -----------------------------

This device requires an evcc sponsorship. Check the following link for what this is and how it works: https://docs.evcc.io/docs/sponsorship

 Are you already a sponsor? Yes
 Please enter the sponsortoken [? for help] ************************************

--------------------------------------------

Please provide the following settings:

 IP address or hostname 192.168.178.11

Testing the configuration of echarge cPH2, cPμ2 ...
   Error: selected device is not a wallbox
```

Einmal war ich schon näher dran:
```
Choose a wallbox: echarge cPH2, cPμ2 
Please provide the following settings: IP address or hostname 192.168.178.11 
Testing the configuration of echarge cPH2, cPμ2 ... 
Error: cannot create charger 'hardybarth-salia': could not change chargemode to manual
```

Beim Starten von evcc:
```
[main  ] INFO 2023/05/15 19:32:28 evcc 0.117.3
[main  ] INFO 2023/05/15 19:32:29 using config file: /etc/evcc.yaml
[main  ] INFO 2023/05/15 19:32:29 starting ui and api at :7070
[db    ] INFO 2023/05/15 19:32:43 using sqlite database: /root/.evcc/evcc.db
[site  ] WARN 2023/05/15 19:32:43 bufferStartSoc must be larger than bufferSoc
[site  ] INFO 2023/05/15 19:32:43 site config:
[site  ] INFO 2023/05/15 19:32:43   meters:      grid ✓ pv ✓ battery ✓
[site  ] INFO 2023/05/15 19:32:43     grid:      power ✓ energy ✓ currents ✗
[site  ] INFO 2023/05/15 19:32:43     pv 1:      power ✓ energy ✗ currents ✗
[site  ] INFO 2023/05/15 19:32:43     battery 1: power ✓ energy ✗ currents ✗ soc ✓ capacity ✗
[lp-1  ] INFO 2023/05/15 19:32:43 loadpoint 1:
[lp-1  ] INFO 2023/05/15 19:32:43   mode:        pv
[lp-1  ] INFO 2023/05/15 19:32:43   charger:     power ✗ energy ✗ currents ✗ phases ✗ wakeup ✗
[lp-1  ] INFO 2023/05/15 19:32:43   meters:      charge ✗
[lp-1  ] DEBUG 2023/05/15 19:32:43 phase timer inactive
[lp-1  ] DEBUG 2023/05/15 19:32:43 pv timer inactive
[lp-1  ] DEBUG 2023/05/15 19:32:43 guard timer inactive
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x14 pc=0x1bbfab4]

goroutine 1 [running]:
github.com/evcc-io/evcc/core.(*Loadpoint).Prepare(0x5900000, 0x642a280, 0x642a2c0, 0x58df180)
	/home/runner/work/evcc/evcc/core/loadpoint.go:609 +0x93c
github.com/evcc-io/evcc/core.(*Site).Prepare(0x6378b40, 0x586d900, 0x5aeac00)
	/home/runner/work/evcc/evcc/core/site.go:766 +0xd8
github.com/evcc-io/evcc/cmd.runRoot(0x3ae9398, {0x3b222b8, 0x0, 0x0})
	/home/runner/work/evcc/evcc/cmd/root.go:260 +0x179c
github.com/spf13/cobra.(*Command).execute(0x3ae9398, {0x580e148, 0x0, 0x0})
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x714
github.com/spf13/cobra.(*Command).ExecuteC(0x3ae9398)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x418
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/evcc-io/evcc/cmd.Execute()
	/home/runner/work/evcc/evcc/cmd/root.go:109 +0x20
main.main()
	/home/runner/work/evcc/evcc/main.go:44 +0x3c
```

Beste Grüße
Tobi
</div>

--- a/charger/hardybarth-salia.go
+++ b/charger/hardybarth-salia.go
@@ -28,6 +28,7 @@ import (
 	"github.com/evcc-io/evcc/api"
 	"github.com/evcc-io/evcc/charger/echarge"
 	"github.com/evcc-io/evcc/charger/echarge/salia"
+	"github.com/evcc-io/evcc/provider"
 	"github.com/evcc-io/evcc/util"
 	"github.com/evcc-io/evcc/util/request"
 	"github.com/evcc-io/evcc/util/sponsor"
@@ -41,9 +42,7 @@ type Salia struct {
 	log     *util.Logger
 	uri     string
 	current int64
-	res     salia.Api
-	cache   time.Duration
-	updated time.Time
+	apiG    provider.Cacheable[salia.Api]
 }
 
 func init() {
@@ -79,18 +78,23 @@ func NewSalia(uri string, cache time.Duration) (api.Charger, error) {
 		Helper:  request.NewHelper(log),
 		uri:     util.DefaultScheme(uri, "http"),
 		current: 6,
-		cache:   cache,
 	}
+
+	wb.apiG = provider.ResettableCached(func() (salia.Api, error) {
+		var res salia.Api
+		err := wb.GetJSON(wb.uri, &res)
+		return res, err
+	}, cache)
 
 	if !sponsor.IsAuthorized() {
 		return nil, api.ErrSponsorRequired
 	}
 
 	// set chargemode manual
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	if err == nil && res.Secc.Port0.Salia.ChargeMode != echarge.ModeManual {
 		if err = wb.post(salia.ChargeMode, echarge.ModeManual); err == nil {
-			res, err = wb.get()
+			res, err = wb.apiG.Get()
 		}
 
 		if err == nil && res.Secc.Port0.Salia.ChargeMode != echarge.ModeManual {
@@ -103,8 +107,7 @@ func NewSalia(uri string, cache time.Duration) (api.Charger, error) {
 
 		wb.pause(false)
 
-		res, err := wb.get()
-		if err == nil && res.Secc.Port0.Metering.Meter.Available > 0 {
+		if res, err := wb.apiG.Get(); err == nil && res.Secc.Port0.Metering.Meter.Available > 0 {
 			return decorateSalia(wb, wb.currentPower, wb.totalEnergy, wb.currents), nil
 		}
 	}
@@ -118,19 +121,6 @@ func (wb *Salia) heartbeat() {
 			wb.log.ERROR.Println("heartbeat:", err)
 		}
 	}
-}
-
-func (wb *Salia) get() (salia.Api, error) {
-	if time.Since(wb.updated) < wb.cache {
-		return wb.res, nil
-	}
-
-	err := wb.GetJSON(wb.uri, &wb.res)
-	if err == nil {
-		wb.updated = time.Now()
-	}
-
-	return wb.res, err
 }
 
 func (wb *Salia) post(key, val string) error {
@@ -150,12 +140,14 @@ func (wb *Salia) post(key, val string) error {
 		}
 	}
 
+	wb.apiG.Reset()
+
 	return err
 }
 
 // Status implements the api.Charger interface
 func (wb *Salia) Status() (api.ChargeStatus, error) {
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	if err != nil {
 		return api.StatusNone, err
 	}
@@ -170,7 +162,7 @@ func (wb *Salia) Status() (api.ChargeStatus, error) {
 
 // Enabled implements the api.Charger interface
 func (wb *Salia) Enabled() (bool, error) {
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	if err == nil && res.Secc.Port0.Salia.ChargeMode != echarge.ModeManual {
 		err = fmt.Errorf("invalid mode: %s", res.Secc.Port0.Salia.ChargeMode)
 	}
@@ -193,19 +185,13 @@ func (wb *Salia) Enable(enable bool) error {
 	err := wb.setCurrent(current)
 	if err == nil {
 		wb.pause(enable)
-		wb.updated = time.Time{}
 	}
 
 	return err
 }
 
 func (wb *Salia) setCurrent(current int64) error {
-	err := wb.post(salia.GridCurrentLimit, strconv.Itoa(int(current)))
-	if err == nil {
-		wb.updated = time.Time{}
-	}
-
-	return err
+	return wb.post(salia.GridCurrentLimit, strconv.Itoa(int(current)))
 }
 
 // MaxCurrent implements the api.Charger interface
@@ -219,19 +205,19 @@ func (wb *Salia) MaxCurrent(current int64) error {
 
 // currentPower implements the api.Meter interface
 func (wb *Salia) currentPower() (float64, error) {
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	return res.Secc.Port0.Metering.Power.ActiveTotal.Actual / 10, err
 }
 
 // totalEnergy implements the api.MeterEnergy interface
 func (wb *Salia) totalEnergy() (float64, error) {
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	return res.Secc.Port0.Metering.Energy.ActiveImport.Actual / 1e3, err
 }
 
 // currents implements the api.PhaseCurrents interface
 func (wb *Salia) currents() (float64, float64, float64, error) {
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	i := res.Secc.Port0.Metering.Current.AC
 	return i.L1.Actual / 1e3, i.L2.Actual / 1e3, i.L3.Actual / 1e3, err
 }
@@ -247,7 +233,7 @@ var _ api.Diagnosis = (*Salia)(nil)
 
 // Diagnose implements the api.Diagnosis interface
 func (wb *Salia) Diagnose() {
-	res, err := wb.get()
+	res, err := wb.apiG.Get()
 	if err == nil {
 		fmt.Printf("Model name: %s\n", res.Device.ModelName)
 		fmt.Printf("Software version: %s\n", res.Device.SoftwareVersion)

--- a/charger/hardybarth-salia.go
+++ b/charger/hardybarth-salia.go
@@ -107,12 +107,12 @@ func NewSalia(uri string, cache time.Duration) (api.Charger, error) {
 
 		wb.pause(false)
 
-		if res, err := wb.apiG.Get(); err == nil && res.Secc.Port0.Metering.Meter.Available > 0 {
+		if res.Secc.Port0.Metering.Meter.Available > 0 {
 			return decorateSalia(wb, wb.currentPower, wb.totalEnergy, wb.currents), nil
 		}
 	}
 
-	return nil, err
+	return wb, err
 }
 
 func (wb *Salia) heartbeat() {

